aten/src/ATen/native/SummaryOps.cpp

```diff
@@ -52,10 +51,77 @@ Tensor _bincount_cpu_template(
   }
   return output;
 }
+
+// A helper function to get the bin for histogram
+// while each bin is inclusive at the lower end and exclusive at the higher,
+// i.e. [start, end) the last bin is inclusive at both, i.e. [start, end], in
+// order to include maxvalue if exists
+template <typename input_t>
+inline int64_t getbin(input_t x, input_t min, input_t max, int64_t bins) {
+  if (x == max)
+    return bins - 1;
+  return (int64_t)((x - min) * bins / (max - min));
+}
+
+///////////////// histogram /////////////////
+template <typename input_t, typename weights_t>
+Tensor _histogram_cpu_template_uniformbins(
+    const Tensor& self,
+    int64_t nbins,
+    const Tensor& weights,
+    input_t min,
+    input_t max,
+    bool density) {
+
+  if (min == max) {
+    min -= 1;
+    max += 1;
+  }
+  TORCH_CHECK(nbins > 0, "bins must be > 0");
+  // Check if range is finite, hack to keep MSVC from complaining
+  TORCH_CHECK(
+      !(std::isinf((double)min) || std::isinf((double)max) ||
+        std::isnan((double)min) || std::isnan((double)max)),
+      "range of [",
+      min,
+      ", ",
+      max,
+      "] is not finite");
+  TORCH_CHECK(min < max, "max must be larger than min");
+
+  bool has_weights = weights.defined();
+  TORCH_CHECK(
+      !has_weights || weights.size(0) == self.size(0),
+      "input and weights should have the same length");
```

Comment:
Having the same number of elements (numel) here doesn't seem like the correct check. Shouldn't it be that weights is broadcastable to the shape of self? (That check might come for free later, too, although maybe not with such a nice error message.)

Reply:
Numpy checks whether weights and self have the same shape, and then flattens the weights into 1D. Although possibly weights being broadcastable to the shape of self would make more sense. Anyway, changing my code so that it checks whether they have the same shape, not only the number of elements.